README.md
<h1 align="center" style="">MxPidQosBuddy</h1>
<p align="center">
Find a PIDs QoS, as well as the range of CPU cores that your system will schedule to host that PID. Available to Apple M-Series.
</p>

<p align="center">
    <a href="">
       <img alt="MacOS" src="https://img.shields.io/badge/MacOS-Apple_Silicon_Only-red.svg"/>
    </a>
    <a href="https://github.com/BitesPotatoBacks/MxPidQosBuddy/releases">
        <img alt="Releases" src="https://img.shields.io/github/release/BitesPotatoBacks/MxPidQosBuddy.svg"/>
    </a>
    <a href="https://github.com/BitesPotatoBacks/MxPidQosBuddy/blob/main/LICENSE">
        <img alt="License" src="https://img.shields.io/github/license/BitesPotatoBacks/MxPidQosBuddy.svg"/>
    </a>
<!--     <a href="https://cash.app/$bitespotatobacks">
        <img alt="License" src="https://img.shields.io/badge/donate-Cash_App-default.svg"/>
    </a> -->
    <br>
</p>

## Project Deets
Based on my tests, the 5 modes of QoS correspond with the scheduling priority of a process. Thus, PID QoS is determined based on scheduling priority (found using `ps` command). The CPU cores available to host the specifed PID(s) are determined based on the activity of each CPU core (sampled using Objective-C) along with the PID QoS.

Unfortunately, Intel Macs are not supported by this project due to the behavior of SMP.

## Installation and Usage
1. Download the .zip file from the [latest release](https://github.com/BitesPotatoBacks/MxPidQosBuddy/releases).
2. Unzip the .zip file and run the `install.sh` script in your terminal, like so: `sudo bash PATH/TO/SCRIPT/install.sh -i`. To see all installer options, use arg `-h`.
3. Once the installation is complete, you may use `./mxpidqosbud`. To see all runtime options, use arg `-h`.

In the output, CPU cores available to host the specifed PID(s) are preseneted in **bold**. Cores out of range will appear in gray.


## Issues
If any bugs or issues are identified or you want your system supported, please let me know in the [issues](https://github.com/BitesPotatoBacks/MxPidQosBuddy/issues) section.

## Support
If you would like to support this project, a small donation to my [Cash App](https://cash.app/$bitespotatobacks) would be much appreciated!



CHANGELOG.md
## v0.2.0 (July 31, 2022)

### Release Notes
This minor includes a crucial bug fixe to the installer, new features, as well as general improvements.

### Changelog

**Features**
- Now the tool can show the task name belonging to the PID
- Added a command line arg `-g` to allow the input of GIDs rather than PIDs

**Bug Fixes**
- Fixed install script not actually adding `/usr/local/bin` to PATH when missing (which could prevent the user from using the tool)

**Improvements**
- Code density improvements
- Renamed project from `MxPidCpuRange` to `MxPidQosBuddy`

___

## v0.2.0 (July 26, 2022)
### Release Notes
Initial release. More features coming soon.


install.sh
#!/bin/sh

#  mxpidaffg.sh
#  MxPidQosBuddy
#
#  Created by BitesPotatoBacks on 6/24/22.
#  Copyright (c) 2022 BitesPotatoBacks. All rights reserved.
#

PROG="mxpidqosbud"
VER="v0.2.0"

RED=$(tput setaf 1)
GREEN=$(tput setaf 2)
PURPLE=$(tput setaf 5)

BOLD=$(tput bold)
NF=$(tput sgr0)

check_sudo() {
    if [ $UID -ne 0 ]; then
        echo "root is required to install/uninstall $PROG, please authenticate"
        exit 1
    fi
}

install() {
    check_sudo
        
    echo " [1/3] Checking PATH"

    if echo $PATH | grep -q "usr/local/bin"; then :; else
        printf "\n\nexport PATH=\"/usr/local/bin:\$PATH\"" >> ~/.zshrc
    fi
    
    echo " [2/3] Preparing scripts"
    chmod 755 $DIR/mxpidqosbud.sh
    cp $DIR/mxpidqosbud.sh $DIR/mxpidqosbud
    xattr -cr $DIR/mxpidqosbud_cpures

    echo " [3/3] Putting things where they need to be"
    sudo mv $DIR/mxpidqosbud /usr/local/bin/mxpidqosbud
    sudo cp $DIR/mxpidqosbud_cpures /usr/local/bin/mxpidqosbud_cpures
    
    echo "${GREEN}Install Complete!${NF}"
}

uninstall() {
    check_sudo
    sudo rm -rf /usr/local/bin/mxpidqosbud
    sudo rm -rf /usr/local/bin/mxpidqosbud_cpures
}

usage() {
    echo "usage: install.sh -h | -V"
    echo "usage: install.sh [-i|-u|-r]"
    echo "Options: "
    echo "   -h   print help and exit"
    echo "   -V   print install version and exit"
    echo "   -i   install $PROG"
    echo "   -u   uninstall $PROG"
    echo "   -r   reinstall $PROG"
    exit
}

[[ $(arch) != "arm64" ]] && echo "${BOLD}$PROG: ${RED}error:${NF} arch unsupported" && exit 1;

while getopts "uirhvV" OPT; do
    case $OPT in
        u) uninstall 2> /dev/null;
           echo "${PURPLE}Uninstalled $PROG ($VER)${NF}";
           exit ;;
        i) echo "${BOLD}Installing $PROG ($VER) ${NF}";
           install 2> /dev/null;
           exit ;;
        r) echo "${BOLD}Reinstalling $PROG ($VER)${NF}";
           uninstall 2> /dev/null;
           echo "${PURPLE}Uninstalled previous version${NF}"
           install 2> /dev/null;
           exit ;;
        v|V) echo "${BOLD}Installing:${NF} $PROG ($VER)" ;;
        h|*) usage;;
    esac
done

[[ $# -eq 0 ]] && usage


src/mxpidcpur.sh
#!/bin/bash

#  mxpidcpur.sh
#  MxPidCpuRange
#
#  Created by BitesPotatoBacks on 6/24/22.
#  Copyright (c) 2022 BitesPotatoBacks. All rights reserved.
#

if [[ $(arch) != "arm64" ]]; then
    echo "${BOLD}$PROG: ${RED}error:${NF} arch $(arch) unsupported"
    exit 1
fi

#
# Variables
#
PROG="mxpidcpur"
VER="v0.1.0"

PCPUS=$(sysctl hw.perflevel0.logicalcpu | awk '{print $2}')
ECPUS=$(sysctl hw.perflevel1.logicalcpu | awk '{print $2}')
CPUS=$(sysctl hw.logicalcpu | awk '{print $2}')

PS_FILE="/Library/Caches/mxpidcpur/ps_out.txt"

BOLD=$(tput bold)
RED=$(tput setaf 1)
NF=$(tput sgr0)

#
# Checking args
#
INTERVAL=56
SHOULD_USE_RES=true

while getopts "i:uhvV" OPT; do
    case $OPT in
        i)   INTERVAL=$OPTARG ;;
        u)   SHOULD_USE_RES=false ;;
        v|V) echo "$PROG $VER"; exit ;;
        h|*) echo "usage: $PROG -h | -V";
             echo "usage: $PROG -u [-i interval] pid[,pid...]";
             echo "Options:";
             echo "   -h     print help and exit";
             echo "   -V     print version and exit";
             echo "   -i<N>  sampling interval for core residency (higher is better) [default:56ms]";
             echo "   -u     disable the use of core residency to narrow the affinity list(s)";
             exit ;;
    esac
done

shift $(($OPTIND - 1))

#
# pulling our pids
#
PIDS=$1

if [[ ! "$PIDS" ]] || [[ $# -gt 1 ]]; then
    echo "${BOLD}$PROG: ${RED}error:${NF} incorrect number of args (recieved $# items)"
    exit 1
fi


PS=$(ps -p $PIDS -o pri= -o pid= > $PS_FILE)

PS_PRI=$(awk '{print $1}' $PS_FILE)
PS_PID=$(awk '{print $2}' $PS_FILE)

PS_PRI=($PS_PRI)
PS_PID=($PS_PID)

#
# loop through all pids and display metrics
#
for (( i = 0; i < ${#PS_PID[@]}; i++)); do

    PRI=$(echo ${PS_PRI[$i]})
    PID=$(echo ${PS_PID[$i]})
    
    QOS=""
    AFFINITY=0
    
    #
    # determine QoS
    #
    if [[ $PRI == 4 ]]; then
        QOS="Background"
    elif [[ $PRI == 20 ]]; then
        QOS="Utility"
    elif [[ $PRI == 31 ]]; then
        QOS="User Initiated/Interactive"
    else
        QOS="Default"
    fi
    
    #
    # find dirty cpu range based on qos
    #
    if [[ $PRI -lt 4 ]] || [[ $PRI == 4 ]]; then
        AFFINITY=ECPUS
    else
        AFFINITY=CPUS
    fi
    
    echo "PID $PID"
    printf "  Available Hosting Cores:  "
    
    # print affinity list with improved range based on core residency metrics (assuming they are enabled)
    for (( ii = 0; ii < $CPUS; ii++ )); do
    
        LABEL="ECPU"
        CURRENT_CORE=$ii
        
        if [[ $CURRENT_CORE -gt $(($ECPUS - 1)) ]]; then
            LABEL="PCPU"
            CURRENT_CORE=$(($ii - $PCPUS))
        fi
        
        if [[ $ii -lt AFFINITY ]];
        then
            if [[ $SHOULD_USE_RES == true ]]; then
                USAGE=$(mxpidcpur_cpures $INTERVAL $CPUS $ii)
            else
                USAGE="100"
            fi
            
            if [[ $USAGE != "0.00" ]]; then
                printf "${BOLD}${LABEL}$CURRENT_CORE ${NF}"
            else
                printf "\033[90m${LABEL}$CURRENT_CORE ${NF}"
            fi
        else
            printf "\033[90m${LABEL}$CURRENT_CORE ${NF}"
        fi
        
    done
    
    # print QoS
    printf "\n  Quality of Service:       $QOS\n\n"
done



src/mxpidcpur_cpures.m
/*
 *  mxpidcpur_cpures.m
 *  MxPidCpuRange
 *
 *  Created by BitesPotatoBacks on 7/25/22.
 *  Copyright (c) 2022 BitesPotatoBacks. All rights reserved.
 */

#include <Foundation/Foundation.h>
#include <mach/processor_info.h>

float getres(int interval, natural_t cpus, int core)
{
    processor_info_array_t sample_a;
    processor_info_array_t sample_b;
    
    mach_msg_type_number_t  proc_msg_cnt;
    
    float total_use   = 0;
    float total_ticks = 0;
    float total_res   = 0;
    
    if (!(host_processor_info(mach_host_self(),
                              PROCESSOR_CPU_LOAD_INFO,
                              &cpus, &sample_a,
                              &proc_msg_cnt) == KERN_SUCCESS))
    { return -1; }
    
    [NSThread sleepForTimeInterval:(interval * 1e-3)];
    
    host_processor_info(mach_host_self(),
                        PROCESSOR_CPU_LOAD_INFO,
                        &cpus,
                        &sample_b,
                        &proc_msg_cnt);
    
    total_use = ((sample_b[(CPU_STATE_MAX * core) + CPU_STATE_USER] - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_USER]) +
            (sample_b[(CPU_STATE_MAX * core) + CPU_STATE_SYSTEM] - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_SYSTEM]) +
            (sample_b[(CPU_STATE_MAX * core) + CPU_STATE_NICE]   - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_NICE]));

    total_ticks = total_use + (sample_b[(CPU_STATE_MAX * core) + CPU_STATE_IDLE] - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_IDLE]);
    
    total_res = ((float) total_use / (float) total_ticks) * 100;
    
    if (isnormal(total_res))
        return total_res;
    else
        return 0;
}

int main(int argc, char * argv[])
{
    natural_t   cpus     = 8;
    int         cpu      = 1;
    int         interval = 256;
    
    if (argc == 4) {
        interval  =             [[NSString stringWithFormat:@"%s", argv[1], nil] intValue];
        cpus      =             [[NSString stringWithFormat:@"%s", argv[2], nil] intValue];
        cpu       = (natural_t) [[NSString stringWithFormat:@"%s", argv[3], nil] intValue];
    } else {
        fprintf(stdout, "incorrect number of args supplied");
        exit(1);
    }
    
    float res = getres(interval, cpus, cpu);
    
    if (res != -1)
        fprintf(stdout, "%.2f\n", res);
    else
        exit(1);
    
    return 0;
}


src/mxpidqosbud_cpures.m
/*
 *  mxpidqosbud_cpures.m
 *  MxPidQosBuddy
 *
 *  Created by BitesPotatoBacks on 7/25/22.
 *  Copyright (c) 2022 BitesPotatoBacks. All rights reserved.
 */

#include <Foundation/Foundation.h>
#include <mach/processor_info.h>

float getres(int interval, natural_t cpus, int core)
{
    processor_info_array_t sample_a;
    processor_info_array_t sample_b;
    
    mach_msg_type_number_t  proc_msg_cnt;
    
    float total_use   = 0;
    float total_ticks = 0;
    float total_res   = 0;
    
    if (!(host_processor_info(mach_host_self(),
                              PROCESSOR_CPU_LOAD_INFO,
                              &cpus, &sample_a,
                              &proc_msg_cnt) == KERN_SUCCESS))
    { return -1; }
    
    [NSThread sleepForTimeInterval:(interval * 1e-3)];
    
    host_processor_info(mach_host_self(),
                        PROCESSOR_CPU_LOAD_INFO,
                        &cpus,
                        &sample_b,
                        &proc_msg_cnt);
    
    total_use = ((sample_b[(CPU_STATE_MAX * core) + CPU_STATE_USER] - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_USER]) +
            (sample_b[(CPU_STATE_MAX * core) + CPU_STATE_SYSTEM] - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_SYSTEM]) +
            (sample_b[(CPU_STATE_MAX * core) + CPU_STATE_NICE]   - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_NICE]));

    total_ticks = total_use + (sample_b[(CPU_STATE_MAX * core) + CPU_STATE_IDLE] - sample_a[(CPU_STATE_MAX * core) + CPU_STATE_IDLE]);
    
    total_res = ((float) total_use / (float) total_ticks) * 100;
    
    if (isnormal(total_res))
        return total_res;
    else
        return 0;
}

int main(int argc, char * argv[])
{
    natural_t   cpus     = 8;
    int         cpu      = 1;
    int         interval = 256;
    
    if (argc == 4) {
        interval  =             [[NSString stringWithFormat:@"%s", argv[1], nil] intValue];
        cpus      =             [[NSString stringWithFormat:@"%s", argv[2], nil] intValue];
        cpu       = (natural_t) [[NSString stringWithFormat:@"%s", argv[3], nil] intValue];
    } else {
        fprintf(stdout, "incorrect number of args supplied");
        exit(1);
    }
    
    float res = getres(interval, cpus, cpu);
    
    if (res != -1)
        fprintf(stdout, "%.2f\n", res);
    else
        exit(1);
    
    return 0;
}


src/mxpidqosbud.sh
#!/bin/bash

#  mxpidqosbud.sh
#  MxPidQosBuddy
#
#  Created by BitesPotatoBacks on 6/24/22.
#  Copyright (c) 2022 BitesPotatoBacks. All rights reserved.
#

PROG="mxpidqosbud"
VER="v0.2.0"
BUILD="Jul 28 2022"

PCPUS=$(sysctl hw.perflevel0.logicalcpu | awk '{print $2}')
ECPUS=$(sysctl hw.perflevel1.logicalcpu | awk '{print $2}')
CPUS=$(sysctl hw.logicalcpu | awk '{print $2}')

BOLD=$(tput bold)
RED=$(tput setaf 1)
NF=$(tput sgr0)

INTERVAL=64
SHOULD_USE_RES=true
SHOULD_USE_GID=false

[[ $(arch) != "arm64" ]] && echo "${BOLD}$PROG: ${RED}error:${NF} arch unsupported" && exit 1;

while getopts "i:guhvV" OPT; do
    case $OPT in
        i)   INTERVAL=$OPTARG ;;
        g)   SHOULD_USE_GID=true ;;
        u)   SHOULD_USE_RES=false ;;
        v|V) echo "$PROG $VER (build $BUILD)";
             exit ;;
        h|*) echo "usage: $PROG -h | -V";
             echo "usage: $PROG -gu [-i interval] pid[,pid...]";
             echo "Options:";
             echo "   -h      print help and exit";
             echo "   -V      print version and exit";
             echo "   -i<N>   sampling interval for core residency (higher is better) [default:64ms]";
             echo "   -g      expect gids rather than pids and return metrics accordingly";
             echo "   -u      disable the use of core residency to narrow the CPU core range(s)";
             exit ;;
    esac
done

shift $(($OPTIND - 1))

if [[ ! "$1" ]] || [[ $# -gt 1 ]]; then
    echo "${BOLD}$PROG: ${RED}error:${NF} incorrect number of args (recieved $# items)"
    exit 1
fi

# Accessing our data from ps command
if [[ $SHOULD_USE_GID == false ]]; then
    PS="ps -p $1 -o pri= -o pid= -o %cpu"
else
    PS="ps -G $1 -o pri= -o pid= -o %cpu"
fi

PS_PRI=($(command $PS | awk '{print $1}'))
PS_PID=($(command $PS | awk '{print $2}'))
PS_USE=($(command $PS | awk '{print $2}'))

# loop through all pids and display metrics
for (( i = 0; i < ${#PS_PID[@]}; i++)); do
    PRI=$(echo ${PS_PRI[$i]})
    PID=$(echo ${PS_PID[$i]})
    PIDUSE=$(echo ${PS_USE[$i]})
    
    QOS="Default"
    AFFINITY=CPUS
    
    # determine QoS
    case $PRI in
        4)  QOS="Background";
            AFFINITY=ECPUS ;; # reflect affinity of QoS 9
        20) QOS="Utility" ;;
        31) QOS="User Initiated/Interactive" ;;
    esac

    basename -- "$(ps -p $PID -o comm=)" | tr -d '\n'; printf " ($PID)\n  Available Hosting Cores:  "
    
    # print affinity list narrowed based on core residency metrics (assuming they are enabled)
    for (( ii = 0; ii < $CPUS; ii++ )); do
        USAGE="100"
        LABEL="ECPU"
        CURRENT_CORE=$ii
        OUTPUT="\033[90m${LABEL}$CURRENT_CORE"
        
        if [[ $CURRENT_CORE -gt $(($ECPUS - 1)) ]]; then
            LABEL="PCPU"
            CURRENT_CORE=$(($ii - $PCPUS))
        fi
        
        if [[ $ii -lt AFFINITY ]]; then
            [[ $SHOULD_USE_RES == true ]] && USAGE=$(./mxpidqosbud_cpures $INTERVAL $CPUS $ii)
            [[ $USAGE != "0.00" ]]        && OUTPUT="${BOLD}${LABEL}$CURRENT_CORE"
        fi
        
        printf "$OUTPUT $NF"
    done
    
    printf "\n  Quality of Service:       $QOS\n\n"
done
